opm/simulators/flow/GenericOutputBlackoilModule.cpp

```diff
@@ -861,84 +865,230 @@ doAllocBuffers(const unsigned bufferSize,
+        // one kw for all fields
+        // RESIDUAL
+        None,
 
-        this->saturation_[phaseIdx].resize(bufferSize, 0.0);
-    }
+        // append GWO to kw name
+        // KR(G|W|O)
+        GWO,
 
-    // And oil pressure
-    fluidPressure_.resize(bufferSize, 0.0);
-    rstKeywords["PRES"] = 0;
-    rstKeywords["PRESSURE"] = 0;
+        // Pure kw controls all, additionally append GWO to kw name for each phase
+        // DEN, DENG, DENW, DENO
+        NGWO,
 
-    if (enableMech_ && eclState_.runspec().mech()) {
-        this->mech_.allocate(bufferSize, rstKeywords);
-    }
+        // append gas/wat/oil to kw name
+        // SGAS, SWAT, SOIL
+        GasWatOil,
 
-    // If TEMP is set in RPTRST we output temperature even if THERMAL
-    // is not activated
-    if (enableEnergy_ || rstKeywords["TEMP"] > 0) {
-        this->temperature_.resize(bufferSize, 0.0);
-        rstKeywords["TEMP"] = 0;
-    }
+        // Pure kw controls all, then first letter of kw and apply gas/wat/oil for each phase.
+        // VISC, VGAS, VWAT, VOIL
+        NGasWatOil,
+    };
 
-    if (FluidSystem::phaseIsActive(oilPhaseIdx)) {
-        rstKeywords["SOIL"] = 0;
-    }
-    if (FluidSystem::phaseIsActive(gasPhaseIdx)) {
-        rstKeywords["SGAS"] = 0;
-    }
-    if (FluidSystem::phaseIsActive(waterPhaseIdx)) {
-        rstKeywords["SWAT"] = 0;
-    }
+    struct Entry {
+        std::variant<ScalarBuffer*,
+                     std::array<ScalarBuffer, numPhases>*,
+                     std::vector<ScalarBuffer*>*> data; // data to resize
+        std::string_view kw; // Field keyword
+        bool cond; // Condition on which to resize field
+        bool optional = false; // True if field keyword need to be in map
+        EntryPhaseType phaseType = EntryPhaseType::GasWatOil; // how phase is encoded in kw
+    };
 
-    if (FluidSystem::enableDissolvedGas()) {
-        rs_.resize(bufferSize, 0.0);
-        rstKeywords["RS"] = 0;
-    }
-    if (FluidSystem::enableDissolvedGasInWater()) {
-        rsw_.resize(bufferSize, 0.0);
-        rstKeywords["RSW"] = 0;
-    }
-    if (FluidSystem::enableVaporizedOil()) {
-        rv_.resize(bufferSize, 0.0);
-        rstKeywords["RV"] = 0;
-    }
-    if (FluidSystem::enableVaporizedWater()) {
-        rvw_.resize(bufferSize, 0.0);
-        rstKeywords["RVW"] = 0;
-    }
+    const auto& oilvap = schedule_[std::max(reportStepNum, 0u)].oilvap();
+    const auto& simConfig = eclState_.getSimulationConfig();
+    using OilVapP = OilVaporizationProperties::OilVaporization;
 
-    if (schedule_[reportStepNum].oilvap().drsdtConvective()) {
-        drsdtcon_.resize(bufferSize, 0.0);
-    }
+    auto pbpd_fields = std::vector{
+        &bubblePointPressure_,
+        &dewPointPressure_
+    };
+    auto rockc_fields = std::vector{
+        &rockCompPorvMultiplier_,
+        &rockCompTransMultiplier_,
+        &swMax_,
+        &minimumOilPressure_,
+        &overburdenPressure_
+    };
```

Comment:
Could these be `std::array<>` instead of `std::vector<>`?

Reply:
sadly no, due to the visitor